aspnetcore/fundamentals/app-state.md

```diff
@@ -193,88 +317,115 @@ app.Use(async (context, next) =>
+For middleware that's only used by a single app, `string` keys are acceptable. Middleware shared between app instances should use unique object keys to avoid key collisions. The following example shows how to use a unique object key defined in a middleware class:
 
-```csharp
-public class SampleMiddleware
-{
-    public static readonly object SampleKey = new Object();
+::: moniker range=">= aspnetcore-2.0"
 
-    public async Task Invoke(HttpContext httpContext)
-    {
-        httpContext.Items[SampleKey] = "some value";
-        // additional code omitted
-    }
-}
-```
+[!code-csharp[](app-state/samples/2.x/SessionSample/Middleware/HttpContextItemsMiddleware.cs?name=snippet1&highlight=4,13)]
+
+::: moniker-end
+
+::: moniker range="< aspnetcore-2.0"
+
+[!code-csharp[](app-state/samples/1.x/SessionSample/Middleware/HttpContextItemsMiddleware.cs?name=snippet1&highlight=5,14)]
+
+::: moniker-end
 
 Other code can access the value stored in `HttpContext.Items` using the key exposed by the middleware class:
 
-```csharp
-public class HomeController : Controller
-{
-    public IActionResult Index()
-    {
-        string value = HttpContext.Items[SampleMiddleware.SampleKey];
-    }
-}
-```
+::: moniker range=">= aspnetcore-2.0"
+
+[!code-csharp[](app-state/samples/2.x/SessionSample/Pages/Index.cshtml.cs?name=snippet3)]
+
+::: moniker-end
+
+::: moniker range="< aspnetcore-2.0"
+
+[!code-csharp[](app-state/samples/1.x/SessionSample/Controllers/HomeController.cs?name=snippet3)]
+
+::: moniker-end
+
+This approach also has the advantage of eliminating the use of key strings in the code.
 
-This approach also has the advantage of eliminating repetition of "magic strings" in multiple places in the code.
+## Cache
+
+Caching is an efficient way to store and retrieve data. The app can control the lifetime of cached items.
+
+Cached data isn't associated with a specific request, user, or session. **Be careful not to cache user-specific data that may be retrieved by other users' requests.**
+
+For more information, see the [Cache responses](xref:performance/caching/index) topic.
 
-## Application state data
+## Dependency Injection
 
 Use [Dependency Injection](xref:fundamentals/dependency-injection) to make data available to all users:
 
-1. Define a service containing the data (for example, a class named `MyAppData`).
+1. Define a service containing the data. For example, a class named `MyAppData` is defined:
 
     ```csharp
     public class MyAppData
     {
-        // Declare properties/methods/etc.
-    } 
+        // Declare properties and methods
+    }
+    ```
+
+2. Add the service class to `Startup.ConfigureServices`:
+
+    ```csharp
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddSingleton<MyAppData>();
+    }
     ```
 
-2. Add the service class to `ConfigureServices` (for example `services.AddSingleton<MyAppData>();`).
+3. Consume the data service class in each page model:
```

Comment:
When the selected version is 1.x, the page model verbiage doesn't make sense. An MVC controller is displayed in that case.